packages/react-reconciler/src/ReactFiberRootScheduler.js

```diff
@@ -0,0 +1,458 @@
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {FiberRoot} from './ReactInternalTypes';
+import type {Lane} from './ReactFiberLane';
+import type {PriorityLevel} from 'scheduler/src/SchedulerPriorities';
+
+import {
+  NoLane,
+  NoLanes,
+  getHighestPriorityLane,
+  getNextLanes,
+  includesOnlyNonUrgentLanes,
+  includesSyncLane,
+  markStarvedLanesAsExpired,
+} from './ReactFiberLane';
+import {
+  CommitContext,
+  NoContext,
+  RenderContext,
+  getExecutionContext,
+  getWorkInProgressRoot,
+  getWorkInProgressRootRenderLanes,
+  isWorkLoopSuspendedOnData,
+  performConcurrentWorkOnRoot,
+  performSyncWorkOnRoot,
+} from './ReactFiberWorkLoop';
+import {LegacyRoot} from './ReactRootTags';
+import {
+  ImmediatePriority as ImmediateSchedulerPriority,
+  UserBlockingPriority as UserBlockingSchedulerPriority,
+  NormalPriority as NormalSchedulerPriority,
+  IdlePriority as IdleSchedulerPriority,
+  cancelCallback as Scheduler_cancelCallback,
+  scheduleCallback as Scheduler_scheduleCallback,
+  now,
+} from './Scheduler';
+import {
+  DiscreteEventPriority,
+  ContinuousEventPriority,
+  DefaultEventPriority,
+  IdleEventPriority,
+  lanesToEventPriority,
+} from './ReactEventPriorities';
+import {supportsMicrotasks, scheduleMicrotask} from './ReactFiberHostConfig';
+
+import ReactSharedInternals from 'shared/ReactSharedInternals';
+const {ReactCurrentActQueue} = ReactSharedInternals;
+
+// A linked list of all the roots with pending work. In an idiomatic app,
+// there's only a single root, but we do support multi root apps, hence this
+// extra complexity. But this module is optimized for the single root case.
+let firstScheduledRoot: FiberRoot | null = null;
+let lastScheduledRoot: FiberRoot | null = null;
+
+// Used to prevent redundant mircotasks from being scheduled.
+let didScheduleMicrotask: boolean = false;
+// `act` "microtasks" are scheduled on the `act` queue instead of an actual
+// microtask, so we have to dedupe those separately. This wouldn't be an issue
+// if we required all `act` calls to be awaited, which we might in the future.
+let didScheduleMicrotask_act: boolean = false;
+
+// Used to quickly bail out of flushSync if there's no sync work to do.
+let mightHavePendingSyncWork: boolean = false;
+
+let isFlushingWork: boolean = false;
+
+export function ensureRootIsScheduled(root: FiberRoot): void {
+  // This function is called whenever a root receives an update. It does two
+  // things 1) it ensures the root is in the root schedule, and 2) it ensures
+  // there's a pending microtask to process the root schedule.
+  //
+  // Most of the actual scheduling logic does not happen until
+  // `scheduleTaskForRootDuringMicrotask` runs.
+
+  // Add the root to the schedule
+  if (root === lastScheduledRoot || root.next !== null) {
+    // Fast path. This root is already scheduled.
```

Comment:
Can it early return here?

Reply:
No because even if a root is scheduled its priority might have changed so we still need to process it in the microtask